File: 802_find_eventual_safe_states.cpp
class Solution {
public:
    vector<int> eventualSafeNodes(vector<vector<int>>& g) {
        int n = g.size();
        vector<bool> vis(n, false), inPath(n, false), safe(n, false);
        for (int i = 0; i < n; ++i) dfs(i, g, vis, inPath, safe);
        vector<int> res;
        for (int i = 0; i < n; ++i) if (safe[i]) res.push_back(i);
        return res;
    }

private:
    bool dfs(int node, vector<vector<int>>& g, vector<bool>& vis, vector<bool>& inPath, vector<bool>& safe) {
        if (vis[node]) return safe[node];
        vis[node] = inPath[node] = true;
        for (int neighbor : g[node]) {
            if (inPath[neighbor] || !dfs(neighbor, g, vis, inPath, safe)) return safe[node] = false;
        }
        inPath[node] = false;
        return safe[node] = true;
    }
};

File: 802_find_eventual_safe_states.java
class Solution {
    public List<Integer> eventualSafeNodes(int[][] graph) {
        int n = graph.length;
        boolean[] visited = new boolean[n];
        boolean[] onPath = new boolean[n];
        boolean[] safe = new boolean[n];

        for (int i = 0; i < n; i++) {
            dfs(i, graph, visited, onPath, safe);
        }

        List<Integer> result = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            if (safe[i]) result.add(i);
        }

        return result;
    }

    private boolean dfs(int node, int[][] graph, boolean[] visited, boolean[] onPath, boolean[] safe) {
        if (visited[node]) return safe[node];
        visited[node] = true;
        onPath[node] = true;

        for (int neighbor : graph[node]) {
            if (onPath[neighbor] || !dfs(neighbor, graph, visited, onPath, safe)) {
                safe[node] = false;
                return safe[node];
            }
        }

        onPath[node] = false;
        safe[node] = true;
        return safe[node];
    }
}
